### PR TITLE
Update vendoring for k8s 1.19

### PR DESCRIPTION
**Pull request by gonzolino**

### CODE REVIEW

**Comment by afritzler** [comment]
/cc @stoyanr

We are currently rewriting the cloud controller to use the controller runtime client + that way bumping the k8s version to 1.19. Still WIP though.

**Review by stoyanr — CHANGES_REQUESTED** [review]
@gonzolino @afritzler Thanks a lot for addressing this! The PR looks good, however when I cloned it locally and tried to build it I encountered some issues, mainly with missing `package` in tests files, as well as with some build errors due to incorrect dependencies. I fixed them locally, you can see the changes I did here: https://github.com/stoyanr/cloud-provider-kubevirt/commit/fa0e55aa8c3f7557577c04c9d3d222295e561ccc. I tried to push this commit to this PR but permission was denied.
Are you going to refactor also the tests? Should I build and test the updated CCM with our kubevirt extension?

**Comment by afritzler** [comment]
This is still WIP.  So there are still glitches. The coding should build now - but we still need to refactor the testing/mocking infrastructure. That is actually a little bit trickier.

**Comment by stoyanr** [comment]
@afritzler The coding doesn't build without the last commit from my fork - as you can see Travis build also fails. You need the commit I pointed above for it to build.
In addition, after I got it to build, I tested it with our kubevirt extension and found some loadbalancer creation issues. See https://github.com/stoyanr/cloud-provider-kubevirt/commit/88c511bf7968e5cbc3caba6e9a5e42eade073285 for the fixes I had to do. With these fixes, the CCM seems to work as expected.
If you give me permissions, I could push these 2 commits to the PR as well. Or you could do the same changes yourselves.

File: go.mod [hunk 1]
```diff
@@ -1,50 +1,81 @@
...
+	github.com/go-logfmt/logfmt v0.5.0 // indirect
+	github.com/golang/mock v1.4.4 // indirect
+	github.com/spf13/cobra v1.1.1
 	github.com/spf13/pflag v1.0.5
-	github.com/xiang90/probing v0.0.0-20190116061207-43a291ad63a2 // indirect
-	gopkg.in/yaml.v2 v2.2.8
-	k8s.io/api v0.17.0
-	k8s.io/apimachinery v0.17.1-beta.0
-	k8s.io/apiserver v0.16.8
+	google.golang.org/grpc/examples v0.0.0-20201021230544-4e8458e5c638 // indirect
```
Comment (stoyanr) [inline go.mod:10]: This should not be here for `make deps-update` to pass
```suggestion
```

File: go.mod [hunk 2]
```diff
@@ -1,50 +1,81 @@
...
-	k8s.io/kubelet => k8s.io/kubelet v0.16.8 // 1.16.8
-	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.16.8 // 1.16.8
-	k8s.io/metrics => k8s.io/metrics v0.16.8 // 1.16.8
-	k8s.io/node-api => k8s.io/node-api v0.16.8 // 1.16.8
-	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.16.8 // 1.16.8
+	github.com/openshift/api => github.com/openshift/api v0.0.0-20191219222812-2987a591a72c
+	github.com/openshift/client-go => github.com/openshift/client-go v0.0.0-20191125132246-f6563a70e19a
+	github.com/operator-framework/operator-registry => github.com/operator-framework/operator-registry v1.14.3
+	google.golang.org/grpc => google.golang.org/grpc v1.27.0
+	k8s.io/api => k8s.io/api v0.19.3
```
Comment (stoyanr) [inline go.mod:29]: You have to pin gopkg.in/square/go-jose.v2 to v2.2.2 or you will get errors during build in k8s deps.
```suggestion
	google.golang.org/grpc => google.golang.org/grpc v1.27.0
	gopkg.in/square/go-jose.v2 => gopkg.in/square/go-jose.v2 v2.2.2
	k8s.io/api => k8s.io/api v0.19.3
```

File: pkg/cloudprovider/kubevirt/cloud_test.go [hunk 3]
```diff
@@ -1,99 +1,99 @@
-package kubevirt
+// package kubevirt
```
Comment (stoyanr) [inline pkg/cloudprovider/kubevirt/cloud_test.go:1]: Uncomment this for the build to pass.
```suggestion
package kubevirt
```

File: pkg/cloudprovider/kubevirt/instances_test.go [hunk 4]
```diff
@@ -1,347 +1,347 @@
...
-		instanceShutdown, err := i.InstanceShutdownByProviderID(context.TODO(), test.providerID)
-		if instanceShutdown != test.expectedInstanceShutdown {
-			t.Errorf("Expected: %v, got: %v", test.expectedInstanceShutdown, instanceShutdown)
-		}
-		if test.expectedError != nil && err != nil && err.Error() != test.expectedError.Error() {
-			t.Errorf("Expected: '%v', got '%v'", test.expectedError, err)
-		}
-	}
-}
+// package kubevirt
```
Comment (stoyanr) [inline pkg/cloudprovider/kubevirt/instances_test.go:1]: Uncomment this for the build to pass.
```suggestion
package kubevirt
```

File: pkg/cloudprovider/kubevirt/loadbalancer.go [hunk 5]
```diff
@@ -167,9 +173,9 @@ func (lb *loadbalancer) EnsureLoadBalancerDeleted(ctx context.Context, clusterNa
 	return nil
 }
 
-func (lb *loadbalancer) getLoadBalancerService(lbName string) (*corev1.Service, bool, error) {
-	service, err := lb.kubevirt.CoreV1().Services(lb.namespace).Get(lbName, metav1.GetOptions{})
-	if err != nil {
+func (lb *loadbalancer) getLoadBalancerService(ctx context.Context, lbName string) (*corev1.Service, bool, error) {
+	var service *corev1.Service
+	if err := lb.client.Get(ctx, client.ObjectKey{Name: lbName, Namespace: lb.namespace}, service); err != nil {
```
Comment (stoyanr) [inline pkg/cloudprovider/kubevirt/loadbalancer.go:178]: You can't pass a nil pointer here,
```suggestion
	var service corev1.Service
	if err := lb.client.Get(ctx, client.ObjectKey{Name: lbName, Namespace: lb.namespace}, &service); err != nil {
```

File: pkg/cloudprovider/kubevirt/loadbalancer.go [hunk 6]
```diff
@@ -178,7 +184,7 @@ func (lb *loadbalancer) getLoadBalancerService(lbName string) (*corev1.Service,
 	return service, true, nil
```
Comment (stoyanr) [inline pkg/cloudprovider/kubevirt/loadbalancer.go:184]:
```suggestion
	return &service, true, nil
```

File: pkg/cloudprovider/kubevirt/loadbalancer.go [hunk 7]
```diff
@@ -223,10 +228,10 @@ func (lb *loadbalancer) createLoadBalancerServicePorts(service *corev1.Service)
...
-func (lb *loadbalancer) applyServiceLabels(lbName, serviceName string, nodes []*corev1.Node) error {
+func (lb *loadbalancer) applyServiceLabels(ctx context.Context, lbName, serviceName string, nodes []*corev1.Node) error {
 	instanceIDs := buildInstanceIDMap(nodes)
-	allVmis, err := lb.kubevirt.VirtualMachineInstance(lb.namespace).List(&metav1.ListOptions{})
-	if err != nil {
+	var allVmis kubevirtv1.VirtualMachineInstanceList
+	if err := lb.client.List(ctx, &allVmis, client.InNamespace(lb.namespace)); err != nil {
 		return fmt.Errorf("Failed to list VMIs: %v", err)
 	}
 	vmiUIDs := make([]string, len(instanceIDs))
```
Comment (stoyanr) [inline pkg/cloudprovider/kubevirt/loadbalancer.go:237]: Don't use make, or you will get a slice with a few empty strings inside, this breaks validation later on.
```suggestion
	var vmiUIDs []string
```

File: pkg/cloudprovider/kubevirt/loadbalancer_test.go [hunk 8]
```diff
@@ -1,1000 +1,1000 @@
-package kubevirt
+// package kubevirt
```
Comment (stoyanr) [inline pkg/cloudprovider/kubevirt/loadbalancer_test.go:1]:
```suggestion
package kubevirt
```

File: pkg/cloudprovider/kubevirt/zones_test.go [hunk 9]
```diff
@@ -1,153 +1,153 @@
...
-		zone, err := z.GetZoneByNodeName(context.TODO(), test.nodeName)
-		if zone != test.expectedZone {
-			t.Errorf("Expected: %v, got: %v", test.expectedZone, zone)
-		}
-		if test.expectedError != nil && err != nil && err.Error() != test.expectedError.Error() {
-			t.Errorf("Expected: '%v', got '%v'", test.expectedError, err)
-		}
-	}
-}
+// package kubevirt
```
Comment (stoyanr) [inline pkg/cloudprovider/kubevirt/zones_test.go:1]:
```suggestion
package kubevirt
```

**Review by stoyanr — CHANGES_REQUESTED** [review]

**Comment by gonzolino** [comment]
> If you give me permissions, I could push these 2 commits to the PR as well. Or you could do the same changes yourselves - I suggested the important ones below.

Thanks @stoyanr ! I cherry-picked your changes into the PR. @afritzler and I will now try to update the unit tests for the new controller-runtime client.

**Comment by stoyanr** [comment]
@gonzolino, I am glad to be able to help, though I still don't understand how do you manage to build anything without the other changes proposed above. In any case, looking forward to the updated PR.

**Review by stoyanr — APPROVED** [review]

**Comment by gonzolino** [comment]
@stoyanr I wanted to merge #14 before finishing this PR. I currently don't have much time to work on this, but I'll try to make some progress over Easter.

**Pull request merged**